multi-node/emulate-multi-node.md

```diff
@@ -92,7 +92,7 @@ The full setup instructions follow:
 
  Script adapted from [torch-distributed-gpu-test.py](./torch-distributed-gpu-test.py) to just tweak `os.environ["CUDA_VISIBLE_DEVICES"]`
  
- ```bash
+```bash
```

Comment:
e.g.

1. Wrapped in \`\`\`bash\`\`\`

    ```bash
    $ cat test2.py
    import deepspeed
    import fcntl
    import os
    import socket
    import time
    import torch
    import torch.distributed as dist
    
    # a critical hack to use the 2nd GPU by the 2nd process (otherwise both processes will use gpu0)
    if os.environ["RANK"] == "1":
       os.environ["CUDA_VISIBLE_DEVICES"] = "1"
    
    def printflock(*msgs):
       """ solves multi-process interleaved print problem """
       with open(__file__, "r") as fh:
           fcntl.flock(fh, fcntl.LOCK_EX)
    ```

2. vs. split as \`\`\`bash\`\`\` and \`\`\`python\`\`\`

    ```bash
    $ cat test2.py
    ```

    ```python
    import deepspeed
    import fcntl
    import os
    import socket
    import time
    import torch
    import torch.distributed as dist
    
    # a critical hack to use the 2nd GPU by the 2nd process (otherwise both processes will use gpu0)
    if os.environ["RANK"] == "1":
       os.environ["CUDA_VISIBLE_DEVICES"] = "1"
    
    def printflock(*msgs):
       """ solves multi-process interleaved print problem """
       with open(__file__, "r") as fh:
           fcntl.flock(fh, fcntl.LOCK_EX)
    ```

Reply:
I hear you, Sam. Again, we use/view the same document in different ways.

And again, I'm not sure what's the best way here.

I know it's inconsistent at the moment but the intention was for the user to quickly create those files in one command w/o opening an editor. except I need to fix it up to be so.

Here is an example of what I meant for it be eventually:

https://github.com/stas00/ml-engineering/blob/fa7424ab6f798c5b6c7494493c7c084cd5d11bbb/slurm/users.md#L583-L590

Thoughts?

Reply:
While we are discussing this it might be a good segue to this relate topic. I started working on making a nice pdf version of this repo, and one of the puzzles is whether to include the code-only files verbatim or leave them as links to the repo.

I'm bring it up since currently it's inconsistent - sometimes it's an inlined code section, sometimes it's a link to an actual script / file.

Just curious if you have an opinion on that matter, it's just related to what we are discussing now.

Reply:
so I'm biased, but I'm a big fan of [Quarto](https://quarto.org/) for converting markdown to nicely formatted outputs.

They have a whole [Gallery](https://quarto.org/docs/gallery/) of different examples and formats, any of which can be converted to one another via [pandoc](https://pandoc.org/) behind the scenes.

A good example of the book format is:

[Modern Polars](https://kevinheavey.github.io/modern-polars/)


and they've got additional details on outputting to PDFs specifically at:

https://quarto.org/docs/output-formats/pdf-basics.html

Reply:
so basically you could write the content in markdown as-is and then do something like

\`\`\`{python}
%%bash
cat << EOT >> test-nodes.py 
#!/usr/bin/env python 
import torch, socket 
print(socket.gethostname(), torch.cuda.is_available()) 
EOT 
chmod a+x ./test-nodes.py 
\`\`\`


which would execute the `cat << EOT [...]` in `bash` as-if it were a Jupyter cell.

see: [Quarto - Execution Options / Shell Commands](https://quarto.org/docs/computations/execution-options.html#shell-commands)

but this might be overkill if you've already got a procedure for generating the PDFs 🤷🏻‍♂️ 

in which case, I'd say it would probably be best to try and include the code-only files verbatim if possible?

One option might be to stick them in an Appendix at the end of the PDF?

that way (with hyperlinks) you could jump back and forth without them taking up too much room in the middle of the text

Reply:
Thank you for the suggestion on Quarto - I haven't heard of it. Your suggestion missed me by one day as I was sorting it out yesterday, I choose to go with the 2 step way - first converting to html and then to pdf, since there are much better tools for dealing with html2pdf.

https://github.com/stas00/ml-engineering/blob/master/utils/md-to-html.py

I still need to work on the style and deal with images before the nice pdf is ready, but the benefit of using html is that I can now run a link checker - the hardest part is to keep the anchor links unbroken. So I've got that working now and found a bunch of broken links and anchors. 

So I will save you suggestions and study them later - if your approach works better I can switch to that one.

Reply:
@stas00 if you're curious, I've done some re-org and tried rendering everything with Quarto in [my fork](https://github.com/saforem2/ml-engineering)

You can see it online at:

https://saforem2.github.io/ml-engineering/

This is mostly just a first (_rough_) pass (and copied + modified from existing Quarto sites I've made),
but I made an effort to fix / change things where appropriate (e.g. links, etc.)

disclaimer: I mostly just did this out of personal curiosity / testing, but thought I'd share incase you're interested 🤷🏻‍♂️

Reply:
as I mentioned in my last reply above I have already sorted out the full pipeline, so unless you find something wrong this is done. Now I'm back working on the content.